testing/src/main/kotlin/org/neo4j/connectors/kafka/testing/assertions/TopicVerifier.kt

```diff
@@ -98,15 +104,38 @@ class TopicVerifier<K, V>(
     try {
       Awaitility.await().atMost(timeout).until {
         consumer.kafkaConsumer.poll(Duration.ofMillis(500)).forEach { receivedMessages.add(it) }
-        val messages = receivedMessages.toList()
-        messages.size == predicates.size &&
-            predicates.foldIndexed(true) { i, prev, predicate ->
-              prev && predicate.test(messages[i])
+        val messages = receivedMessages.toList().toMutableList()
+
+        if (messages.size != predicates.size) {
+          return@until false
+        }
+
+        if (inAnyOrder) {
+          var queue = predicates.toList()
```

Comment:
wondering if the whole if block could not be replaced with:
```kotlin
predicates.fold(true) { prev, predicate ->
    prev && 
        messages.fold(false) { matched, message } ->
            matched || predicate.test(message)
}
```
I've not tested it though, and it might be slightly less readable :P

Reply:
oh, except, you need to remove the matching message from `messages`...
```kotlin
```kotlin
predicates.fold(true) { prev, predicate ->
    prev && 
        messages.fold(false) { matched, message } -> {
            if (matched) {
                return true;
            }
            if (predicate.test(message)) {
                messages.remove(message); // hope it does not raise a concurrent modification exception
                return true;
            }
            return false;
        }
}
```

Reply:
not sure there's much to again from my suggestions then (assuming they even work)

Reply:
will leave as it is, as it doesn't improve readability as you highlighted.